src/Symfony/Component/HttpKernel/Controller/ErrorController.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller;
+
+use Symfony\Component\ErrorRenderer\ErrorRenderer;
+use Symfony\Component\ErrorRenderer\Exception\ErrorRendererNotFoundException;
+use Symfony\Component\ErrorRenderer\Exception\FlattenException;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+
+/**
+ * Renders error or exception pages for a given FlattenException.
+ *
+ * @author Yonel Ceruto <yonelceruto@gmail.com>
+ * @author Matthias Pigulla <mp@webfactory.de>
+ */
+class ErrorController
+{
+    private $kernel;
+    private $controller;
+    private $errorRenderer;
+
+    public function __construct(HttpKernelInterface $kernel, $controller, ErrorRenderer $errorRenderer)
+    {
+        $this->kernel = $kernel;
+        $this->controller = $controller;
+        $this->errorRenderer = $errorRenderer;
+    }
+
+    public function __invoke(Request $request, FlattenException $exception): Response
+    {
+        try {
+            return new Response($this->errorRenderer->render($exception, $request->getPreferredFormat()), $exception->getStatusCode(), $exception->getHeaders());
+        } catch (ErrorRendererNotFoundException $e) {
+            return new Response($this->errorRenderer->render($exception), $exception->getStatusCode(), $exception->getHeaders());
+        }
+    }
+
+    public function preview(Request $request, int $code): Response
+    {
+        $exception = FlattenException::createFromThrowable(new \Exception('This is a sample exception.'), $code, ['X-Debug' => false]);
+
+        /*
+         * This Request mimics the parameters set by
+         * \Symfony\Component\HttpKernel\EventListener\ExceptionListener::duplicateRequest, with
+         * the additional "showException" flag.
+         */
+        $subRequest = $request->duplicate(null, null, [
+            '_controller' => $this->controller,
+            'exception' => $exception,
+            'logger' => null,
+            'showException' => false,
+        ]);
+
+        return $this->kernel->handle($subRequest, HttpKernelInterface::SUB_REQUEST);
```

Comment:
As I said several times like in https://github.com/symfony/symfony/pull/33065#discussion_r312337877. IMO we should get rid of sub requests for exceptions. This would also change this PR heavily.

Reply:
Using the error rendering tool directly here would leave out a custom exception controller. Is your plan remove that layer/feature "custom exception controller" ?

Reply:
I think a custom exception controller configuration could be fine. But in theory it's not needed. There is a direct solution without any need of a configuration or indirection with an "exception controller": Just let people define their own exception listener with a higher priority than the one provided by the framework bundle. In this listener they can do anything, exactly like the exception controller.
But having an exception controller config is also fine. In the end it's basically the same as the exception listener would just call the (custom) exception controller.

What I want to avoid is the sub request. The exception controller is just a callable. Instead of creating a sub request with the controller, just call the controller directly. The problem with the subrequest is that it triggers all the kernel events again. This has been painful ever since. If you have an error in a listener -> it triggers the exception event -> it triggers the exception listener -> it creates a subrequest -> kernel events trigger again (so likely your failing listener as well) -> the same error happens again -> symfony just displays an error occured while handling an exception.

Reply:
> I think a custom exception controller configuration could be fine. But in theory it's not needed. There is a direct solution without any need of a configuration or indirection with an "exception controller": Just let people define their own exception listener with a higher priority than the one provided by the framework bundle. In this listener they can do anything, exactly like the exception controller.
But having an exception controller config is also fine. In the end it's basically the same as the exception listener would just call the (custom) exception controller.

Sure, but IMO it's not the safest and easiest way to get it. Just look at this example in the doc: https://symfony.com/doc/current/event_dispatcher.html#creating-an-event-listener vs this:
```php
class CustomExceptionController 
{
    public function __invoke(Request $request, FlattenException $exception) 
    {
        return new Response(
            sprintf('My Error says: %s with code: %s', $exception->getMessage(), $exception->getStatusCode()),
            $exception->getStatusCode(),
            $exception->getHeaders()
        );
    }
}
```
Also "Controller" is a basic concept that everyone know just starting with Symfony, In the other hand, "Listener" and exception listeners come with more concepts: events, priority, `$event->getException()` vs `$exception` param, work to send the correct status code, `->setResponse()` vs `return new Response()`, raw `\Throwable` vs `FlattenException`, etc. 

That's why I think promoting exception controller is the way to go here.

> What I want to avoid is the sub request. The exception controller is just a callable. Instead of creating a sub request with the controller, just call the controller directly. 

I got your point, but then we wouldn't be talking about a controller, but a callable with strict parameters, and we would lose many of the great features available today for controllers: DI, automatic wiring, etc. That doesn't sound good to me :/

> The problem with the subrequest is that it triggers all the kernel events again. This has been painful ever since. If you have an error in a listener -> it triggers the exception event -> it triggers the exception listener -> it creates a subrequest -> kernel events trigger again (so likely your failing listener as well) -> the same error happens again -> symfony just displays an error occured while handling an exception.

I agree that it's painful triggers all kernel events again.

Reply:
>lose many of the great features available today for controllers: DI, automatic wiring

A callable can be a service and with that you have all the same stuff like autowiring etc.
Also the controller only works with http errors. If you want to handle error responses for the CLI, you still need the event listener approach.

> $event->getException() vs $exception param

$event->getException() is much better to me because it's typehinted and discoverable. How do I know I need to accept a `FlattenException` in the controller? Normally I know the params because they are based on the route I create. In this case it's arbitrary and not discoverable. So people need to know how an exception controller has to look like.

Reply:
I think the best thing would be to create an alternative "PR" with your idea, could you please do that? I don't see how to remove the sub-application without making this process more complex.

Reply:
I will try to finish this by keeping the controller's focus, so that we can decide later if it is worth it as it is or if we can do it later. Do you agree with that?